### PR TITLE
fixes for --read-special mode

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
please review if the semantics is now right.

### CODE REVIEW

**Comment by PlasmaPower** [comment]
Can we reliably test this?

File: borg/archiver.py [hunk 1]
```diff
@@ -304,10 +304,16 @@ def _process(self, archive, cache, matcher, exclude_caches, exclude_if_present,
                 status = archive.process_symlink(path, st)
         elif stat.S_ISFIFO(st.st_mode):
             if not dry_run:
-                status = archive.process_fifo(path, st)
+                if not read_special:
+                    status = archive.process_fifo(path, st)
+                else:
+                    status = archive.process_file(path, st, cache)
```
Comment (textshell) [inline borg/archiver.py]: st.st_mode still says this is a special file. Should that be masked out?

File: borg/archiver.py [hunk 2]
```diff
@@ -304,10 +304,16 @@ def _process(self, archive, cache, matcher, exclude_caches, exclude_if_present,
...
         elif stat.S_ISCHR(st.st_mode) or stat.S_ISBLK(st.st_mode):
             if not dry_run:
-                status = archive.process_dev(path, st)
+                if not read_special:
+                    status = archive.process_dev(path, st)
+                else:
+                    status = archive.process_file(path, st, cache)
```
Comment (textshell) [inline borg/archiver.py]: st.st_mode still says this is a special file. Should that be masked out?

Comment (enkore) [inline borg/archiver.py]: process_file could set the mode bits of the item to S_IFREG unconditionally.

**Comment by textshell** [comment]
So other then your TODOs this looks good.

about the symlinks. For many use cases in /dev following symlinks for block special devices is needed i’m not even sure udev/kernel commit to what is a symlink and what the real device node as abi. On the other hand, --read-special is not a good option for follow all symlinks.

**Comment by ThomasWaldmann** [comment]
mode should be ok now. i amended the previous commit with a cleaner implementation of that.

**Comment by ThomasWaldmann** [comment]
ok, added symlink handling.

**Comment by textshell** [comment]
I think we should spell out what --read-special means in the documentation. It certainly not guessable anymore. For the implementation i think we can hardly restrict this anymore without breaking usages.

>       --read-special        open and read block and character special files 
                            and FIFOs as if they were regular files. 
                            Also follows symlink pointing to these kinds files.

**Comment by ThomasWaldmann** [comment]
collapse / merge?

**Comment by textshell** [comment]
seems ready to merge.

**Comment by enkore** [comment]
lgtm, rebase/merge

**Pull request merged**